Vibrational problem

dlasecki opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Vibrational problem implementation


### Details and comments

mrossinek commented on qiskit_nature/problems/second_quantization/base_problem.py:
```diff
@@ -0,0 +1,59 @@
+            driver: A driver encoding the molecule information.
+            transformers: A list of transformations to be applied to the molecule.
+        """
+        if transformers is None:
+            transformers = []
+        self.driver = driver
+        self.transformers = transformers
```
We can also do the following but I am not opinionated about this:
```suggestion
        self.driver = driver
        self.transformers = transformers or []
```

dlasecki replied:
Your way is much better. Thanks.

mrossinek commented on qiskit_nature/problems/second_quantization/base_problem.py:
```diff
@@ -0,0 +1,59 @@
+        Returns:
+            A list of `SecondQuantizedOp` in the following order: ... .
+        """
+        return
```
```suggestion
        raise NotImplementedError
```

mrossinek commented on qiskit_nature/problems/second_quantization/base_problem.py:
```diff
@@ -0,0 +1,59 @@
+
+    @abstractmethod
+    def _transform(self, molecule):
+        return
```
```suggestion
        raise NotImplementedError
```

mrossinek commented on qiskit_nature/problems/second_quantization/vibrational/spin_op_builder.py:
```diff
@@ -0,0 +1,68 @@
+        truncation_order: order at which an n-body expansion is truncated
+
+    Returns:
+        SpinOp: SpinOp built from a QMolecule object.
```
```suggestion
        SpinOp: SpinOp built from a `WatsonHamiltonian`.
```

mrossinek commented on qiskit_nature/problems/second_quantization/vibrational/spin_op_builder.py:
```diff
@@ -0,0 +1,68 @@
+
+    num_modes = watson_hamiltonian.num_modes
+    basis_size = [basis_size] * num_modes
+    boson_hamilt_harm_basis = HarmonicBasis(watson_hamiltonian,  # type: ignore
```
I think we should make `HarmonicBasis` configurable and instead take any `BosonicBasis` as an input somewhere.

dlasecki replied:
Sure. Let's deal with it a bit later. Currently we only have a `HarmonicBasis` implemented.

mrossinek commented on qiskit_nature/problems/second_quantization/vibrational/vibrational_problem.py:
```diff
@@ -0,0 +1,56 @@
+        """
+        watson_hamiltonian = self.driver.run()
+        watson_hamiltonian_transformed = self._transform(watson_hamiltonian)
+        basis_size = 1  # TODO how to get it?
+        truncation_order = 3  # TODO how to get it?
```
Afaict from the old way of using `BosonicTransformation` (https://github.com/Qiskit/qiskit-aqua/blob/master/test/chemistry/test_bosonic_esc_calculation.py#L60-L64) this will need to be an input to the constructor. @paulineollitrault Please correct me, if these can be inferred in some other way.

dlasecki replied:
I have been inclined towards putting it to the `VibrationalProblem` constructor. Another option would be arguments of `second_q_ops` but it conflicts with a signature of that method in a parent class.

mrossinek replied:
No `second_q_ops` should not have any arguments. Once you call that, everything should really be set.
The idea here is that it should feel like a property that you query

dlasecki commented on qiskit_nature/algorithms/ground_state_solvers/orbital_optimization_vqe.py:
```diff
@@ -208,8 +208,8 @@ def _initialize_additional_parameters(self, driver: BaseDriver):
             raise QiskitNatureError(
-                            'Please set iterative_oo_iterations parameter to a positive number,'
-                            ' got {} instead'.format(self._iterative_oo_iterations))
+                'Please set iterative_oo_iterations parameter to a positive number,'
```
Lint was complaining a lot.

mrossinek replied:
Even in the CI?

dlasecki commented:
Please take a look at the current code. Updated SpinOp and relevant tests were taken from the PR #68.
 
Tests are passing but there are still some mypy errors:

```
qiskit_nature/operators/second_quantization/spin_op.py:219: error: Incompatible types in assignment (expression has type "List[Tuple[str, int]]", variable has type "Union[str, List[Tuple[str, complex]], List[Tuple[str, float]], Tuple[Any, Any]]")
qiskit_nature/problems/second_quantization/vibrational/vibrational_problem.py:46: error: "BaseDriver" has no attribute "run"
qiskit_nature/problems/second_quantization/molecular/molecular_problem.py:50: error: "BaseDriver" has no attribute "run"
```

Last 2 errors come from the fact of me introducing the class `BaseProblem` that operates on `BaseDriver` that do not have a run method. `BaseDriver` problem has been discussed with @mrossinek and we decided to postpone it. Perhaps we should also postpone introducing `BaseProblem`, what do you think?

I am not sure about the first mypy error. @ikkoham ? I introduced an additional input type that uses float coefficients instead of `complex` coefficients because `VibrationalSpinOp` operates on `float` coefficient only as of now (it starts from the `HarmonicBasis` class). Somehow `float `is not considered a subtype of `complex` in type hints...

ikkoham commented on qiskit_nature/operators/second_quantization/spin_op.py:
```diff
@@ -166,27 +170,23 @@ class SpinOp(SecondQuantizedOp):
+            data: Union[
+                str,
+                List[Tuple[str, complex]],
+                List[Tuple[str, float]],
```
This line is unnecessary because float is a virtual subclass of complex.

mrossinek commented on qiskit_nature/problems/second_quantization/vibrational/vibrational_labels_validator.py:
```diff
@@ -0,0 +1,120 @@
+
+            last_op, last_mode_index, last_modal_index = op, mode_index, modal_index
+
+            increment = 1 if op == "+" else -1
+            check_list[int(mode_index)] += increment
```
```suggestion
            check_list[int(mode_index)] += 1 if op == "+" else -1
```

mrossinek commented on qiskit_nature/problems/second_quantization/vibrational/vibrational_labels_validator.py:
```diff
@@ -0,0 +1,120 @@
+                      num_modals: List[int]):
+    for labels, _ in vibrational_labels:
+        coeff_labels_split = labels.split(" ")
+        check_list = [0] * num_modes
```
Either a better name or a comment explaining what this variable is used for would be great :+1:

mrossinek commented on qiskit_nature/problems/second_quantization/vibrational/vibrational_labels_validator.py:
```diff
@@ -0,0 +1,120 @@
+            check_list[int(mode_index)] += increment
+        if not all(v == 0 for v in check_list):
+            raise ValueError(
+                f"Modes of raising and lowering operators do not agree for labels {labels}.")
```
You could even use the `check_list` to only print the violating labels, right?

mrossinek commented on qiskit_nature/problems/second_quantization/vibrational/vibrational_labels_validator.py:
```diff
@@ -0,0 +1,120 @@
+
+def _is_order_incorrect(mode_index: int, last_mode_index: int, modal_index: int,
+                        last_modal_index: int, op: str, last_op: str) -> bool:
+    return _is_mode_order_incorrect(mode_index, last_mode_index) or \
+           _is_modal_order_incorrect(last_mode_index, mode_index, last_modal_index, modal_index) \
+           or _is_operator_order_incorrect(mode_index, last_mode_index, modal_index,
+                                           last_modal_index, op, last_op)
```
I think you can inline those methods with a comment explaining each rather than splitting out into extra functions. But here I don't mind.

mrossinek commented on qiskit_nature/problems/second_quantization/vibrational/vibrational_spin_op_builder.py:
```diff
@@ -0,0 +1,47 @@
+    num_modes = watson_hamiltonian.num_modes
+    if isinstance(basis_size, int):
+        basis_size = [basis_size] * num_modes
+    # TODO make HarmonicBasis an argument and support other bases when implemented
```
Do you want to address this before merging or in the future?

dlasecki replied:
In the future. Would be better to see how the non-existing alternative to HarmonicBasis looks like and works.

mrossinek commented on qiskit_nature/problems/second_quantization/vibrational/vibrational_spin_op_builder.py:
```diff
@@ -0,0 +1,47 @@
+        basis_size = [basis_size] * num_modes
+    # TODO make HarmonicBasis an argument and support other bases when implemented
+    boson_hamilt_harm_basis = HarmonicBasis(watson_hamiltonian,
+                                            # type: ignore
```
Why is this necessary?

dlasecki replied:
Do you mean the ignore? Not sure, it is a copy-paste from a deprecated class.

mrossinek replied:
Yes. This comment disables `mypy` for the variable in question but I cannot see which variable it is referring to here. Please remove it and see what breaks. Then you can add it back in (if necessary) while making sure it is clear which variable it tells `mypy` to ignore.

dlasecki replied:
No complaints from mypy or lint when I got rid of it. Seems unnecessary.

mrossinek commented on qiskit_nature/transformers/base_transformer.py:
```diff
@@ -18,14 +18,15 @@
     """
 
     @abstractmethod
     def transform(self, q_molecule: QMolecule) -> QMolecule:
```
Since you want to add the `WatsonHamiltonian` to this interface, it should also go in the type hint here. Although that will likely lead to `mypy` issues, too..

dlasecki replied:
That was my initial try but mypy issues, exactly.

mrossinek replied:
You could add a `# type: ignore` comment in these cases just to get `mypy` to shut up for now, until we refactor the drivers and `QMolecule`/`WatsonHamiltonian` classes.

mrossinek commented on test/problems/second_quantization/vibrational/test_vibr_to_spin_op_label_converter.py:
```diff
@@ -0,0 +1,73 @@
+        expected_spin_op_labels = [('-_1 +_0', 1215.375), ('-_5 +_4 -_1 +_0', -6.385)]
+        vibrational_labels = [('-_0*1 +_0*0', 1215.375), ('-_2*1 +_2*0 -_0*1 +_0*0', -6.385)]
+        num_modes = 3
+        num_modals = 2
```
I think you can use `@data` here to reduce some of the code duplication

mrossinek commented on test/problems/second_quantization/vibrational/test_vibr_to_spin_op_label_converter.py:
```diff
@@ -0,0 +1,73 @@
+
+    def test_convert_to_spin_op_labels(self):
+        """Tests that VibrationalSpinOp labels are converted to SpinOp labels correctly."""
+        expected_spin_op_labels = [('-_1 +_0', 1215.375), ('-_5 +_4 -_1 +_0', -6.385)]
```
Where do these random numbers come from?

dlasecki replied:
I copied them from an output for a particle I loaded. Not sure how to explain it here... They don't play any role in these tests, could be anything.

mrossinek commented on test/problems/second_quantization/vibrational/test_vibrational_label_builder.py:
```diff
@@ -0,0 +1,163 @@
+        logfile = self.get_resource_path('CO2_freq_B3LYP_ccpVDZ.log')
+        driver = GaussianForcesDriver(logfile=logfile)
+
+        # TODO extract somewhere
```
Please do :wink:

mrossinek reviewed (changes requested):
Nice work @dlasecki! :+1: 
As usual some nit-picking comments from my side :wink:

dlasecki commented on qiskit_nature/operators/second_quantization/vibrational_spin_op.py:
```diff
@@ -0,0 +1,96 @@
+from typing import List, Tuple, Union
+
+from .. import SpinOp
+from ...problems.second_quantization.vibrational.vibr_to_spin_op_label_converter import \
+    convert_to_spin_op_labels
+from ...problems.second_quantization.vibrational.vibrational_labels_validator import \
+    validate_vibrational_labels
```
We could also move these classes to where the operator sits. I think I did not have a good reason for keeping them in the problem module.

dlasecki replied:
I moved things around a little.

dlasecki commented on qiskit_nature/problems/second_quantization/vibrational/vibr_to_spin_op_label_converter.py:
```diff
@@ -0,0 +1,86 @@
+
+
+def _build_coeff_spin_op_labels(labels: str, partial_sum_modals: List[int]) -> str:
+    coeff_labels_split = labels.split(" ")
```
Yep

dlasecki commented on qiskit_nature/problems/second_quantization/vibrational/vibrational_labels_validator.py:
```diff
@@ -0,0 +1,120 @@
+        Raises:
+            ValueError: if invalid vibrational labels provided.
+        """
+    if isinstance(num_modals, int):
+        num_modals = [num_modals] * num_modes
```
I assumed there might be several entry points.

dlasecki commented on qiskit_nature/problems/second_quantization/vibrational/vibrational_labels_validator.py:
```diff
@@ -0,0 +1,120 @@
+    for labels, _ in vibrational_labels:
+        coeff_labels_split = labels.split(" ")
+        check_list = [0] * num_modes
+        last_op, last_mode_index, last_modal_index = "+", num_modes, max(num_modals)
```
Makes sense.

dlasecki commented on qiskit_nature/problems/second_quantization/vibrational/vibrational_labels_validator.py:
```diff
@@ -0,0 +1,120 @@
+
+def _is_label_duplicated(mode_index: int, last_mode_index: int, modal_index: int,
+                         last_modal_index: int, op: str, last_op: str) -> bool:
+    return modal_index == last_modal_index and mode_index == last_mode_index and op == last_op
```
I think it is sufficient.
Case 1: several duplicated labels next to each other -> we get an error until we deduplicate to a single one,
Case 2: several duplicated labels not next to each other -> we get an ordering error.

mrossinek replied:
Please add a unittest which covers this case though :+1:

dlasecki commented on qiskit_nature/problems/second_quantization/vibrational/vibr_to_spin_op_label_converter.py:
```diff
@@ -0,0 +1,86 @@
+
+    validate_vibrational_labels(vibrational_labels, num_modes, num_modals)
+
+    partial_sum_modals = _calc_partial_sum_modals(num_modals)
```
Good suggestion, thanks,

dlasecki commented on qiskit_nature/problems/second_quantization/vibrational/vibr_to_spin_op_label_converter.py:
```diff
@@ -0,0 +1,86 @@
+    return new_label
+
+
+def _get_ind_from_mode_modal(partial_sum_modals: List[int], mode_index: int,
```
I think it improves readability.

dlasecki commented on qiskit_nature/problems/second_quantization/vibrational/vibrational_label_builder.py:
```diff
@@ -0,0 +1,57 @@
+        A list of labels and corresponding coefficients that describe a vibrational problem.
+    """
+    all_labels = []
+    for num_body in range(truncation_order):
+        num_body_data = boson_hamilt_harm_basis[num_body]
```
I think you are right.

woodsp-ibm commented on qiskit_nature/drivers/base_driver.py:
```diff
@@ -50,6 +50,13 @@ def __init__(self,
         self._hf_method = hf_method
         self._supports_molecule = supports_molecule
 
+    @abstractmethod
```
I imagine the reason we never put a run() method here before is that the type returned from the two subclasses is different and hence we left the run methods to be specific for the fermionic and bosonic drivers but did not try to have some common parent. If this is needed for some reason then maybe it should return 'Any' ?

dlasecki replied:
Yes, this is indeed the case. I introduced a BaseProblem class and it makes sense that it operates on a BaseDriver (it happens that it also runs a driver). I could try 'Any'. We also discussed with Max that these classes will likely be refactored later somehow and we will try to come up with a better solution.

woodsp-ibm commented on qiskit_nature/operators/second_quantization/vibrational_spin_op.py:
```diff
@@ -0,0 +1,102 @@
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""A Vibrational Spin operator.
```
I am not sure that this Note will appear in the Sphinx. I know the class docstring an init docstring get merged. Perhaps build the html and check. Maybe its safer to have this note in the class docstring below (one starting at line 30)

woodsp-ibm commented on qiskit_nature/operators/second_quantization/vibrational_spin_op_utils/vibr_to_spin_op_label_converter.py:
```diff
@@ -0,0 +1,77 @@
+    validate_vibrational_labels
+
+
+def convert_to_spin_op_labels(vibrational_labels: List[Tuple[str, complex]], num_modes: int,
```
This is some separate module, and public supported API, outside of the VibrationalSpinOp class as its considered useful in its own right. I could see this as function in the VibrationalSpinOp - just wondering why we need this and the others in this folder that seem more like private helper functions of that class

woodsp-ibm replied:
Update, I see there was a conversation above about this - sorry I have been trying to catch up with things and did not see that first!

dlasecki replied:
I can see your point. My main motivation was to extract logic with a certain responsibility from the main class. As you can see these modules end up being pretty lengthy. Then, I have been wondering if they could be useful to a user in their own right in case they do some steps on their own - but I am not convinced about it. I would be fine with making all methods in these modules "private" such that they are not exposed as a public API. Let me know if there is a better way to accomplish it and avoid importing a private method to VibrationalSpinOp.

woodsp-ibm replied:
I guess personally I am less worried about the resultant total length of VibrationalSpinOp if these are added at the end of it. The methods themselves are reasonable and so reading them in the overall file should be fine. It would only end up being 2 ro 3 hundred lines long right - puny compared to say the 2.6K that is QuantumCircuit!

woodsp-ibm commented on qiskit_nature/problems/second_quantization/molecular/molecular_problem.py:
```diff
@@ -22,9 +22,10 @@
+class MolecularProblem(BaseProblem):
     """Molecular Problem"""
 
     def __init__(self, fermionic_driver: FermionicDriver,
```
In the same way there is FermionicDriver in here I wonder whether we ought to have FermionicTransformer and likewise BosonicTransformer (here it would be a List of Fermionic ones of course)

mrossinek replied:
Why is this field even called `fermionic_driver` rather than just `driver` as in `BaseProblem`? Sub-typing it should be fine

woodsp-ibm commented on qiskit_nature/transformers/base_transformer.py:
```diff
@@ -13,24 +13,27 @@
-    def transform(self, q_molecule: QMolecule) -> QMolecule:
-        """Transforms one `QMolecule` into another one. This may or may not affect the size of the
-        Hilbert space.
+    def transform(self, molecule_data: Union[QMolecule, WatsonHamiltonian]):
```
Will we ever have transformers that would work on either type. Or more like my comment above would it be better to define A FermionicTransform taking QMolecule and a BosonicTransform taking the WatsonHamiltonian

dlasecki replied:
Not sure about it... @mrossinek ? If not, I agree with you.

mrossinek replied:
I don't think we will ever have transformers which support both types. Maybe you can also just specify `Any` as the type here and then subclass to a more specific type in those subclasses.

I would even argue that, since we do not even have any transformer for `WatsonHamiltonian` we simply leave that out for this release..

dlasecki commented on qiskit_nature/transformers/freeze_core_transformer.py:
```diff
@@ -12,15 +12,14 @@
     """The Freeze-Core transformer."""
 
-    def transform(self, q_molecule: QMolecule) -> QMolecule:
+    def transform(self, q_molecule: QMolecule) -> QMolecule:  # type: ignore
```
Yes

dlasecki commented on qiskit_nature/operators/second_quantization/vibrational_spin_op.py:
```diff
@@ -0,0 +1,96 @@
+    operators together with indices of modes and modals that they act on, e.g. "+_{mode_index}*{
+    modal_index}". Each modal can be excited at most once.
+    **Initialization**
+    # TODO
```
Me neither, do you have anything specific in mind? It is mostly based on the SpinOp docstring.

woodsp-ibm replied:
You can always `make html` locally which uses Sphinx to build things out then you can see how things turn out and make any adjustments

dlasecki commented on qiskit_nature/problems/second_quantization/vibrational/vibrational_labels_validator.py:
```diff
@@ -0,0 +1,120 @@
+
+
+def _validate_data_type(vibrational_labels: List[Tuple[str, float]]):
+    if not isinstance(vibrational_labels, list):
```
Hmm, probably going deeper would be the role of SpinOp since we here we have the same data structure as there.

pbark reviewed (approved):
LGTM.! Thanks @dlasecki 👍
